### ISSUE TITLE
angular templates must have exactly one root element

### ISSUE THREAD

**Issue by jhanink** [cypress-io/cypress#257]
Cypress injects <head><script>...</script><head> to html resources to set document.domain, but angular $compile, directly on that fragment would fail, due to this requirement:

https://docs.angularjs.org/error/$compile/tplrt

A possible workaround would be to detect that the template is an angular template. Maybe a configuration option to set a path (for which no injection is applied).

Another possibility is to wrap the entire construct (injection plus template) in an outer enclosing div... but that has the possibility of breaking the app as well if it relies on a precise DOM structure.

Perhaps a config that gives the choice of these two options?

Here is a copy of the modified template response (bearing the injected header)

``` html
<head> <script type='text/javascript'> document.domain = 'localhost'; </script> </head><div class="pl-table">

  <!-- table filters  -->
  <div class="pl-table-filters" ng-if="config.getIsEnabledFilters()">
    <div ng-show="config.getIsEnabledFilters() && (config.getIsShowFilters() || !isInitialFilterState())">
      <span class="pl-table-filtering-n-results">Total: <span class="bold">{{config.getTableState().unfilteredTotalRecordCount | number}}</span>
        <span ng-show="config.getTableState().unfilteredTotalRecordCount !== table.page.info().recordsDisplay"><span class="bold"> | {{table.page.info().recordsDisplay | number}}</span> Filtered</span></span>
      <button class="pl-button bare pl-table-clear-filters" ng-disabled="isInitialFilterState()" ng-class="{'disabled': isInitialFilterState()}" ng-click="clearFilters()">Clear All Filters</button>
    </div>
    <div ng-show="config.getIsEnabledFilters() && config.getIsShowFilters()">
      <div class="pl-table-filter-criteria-group">
        <div class="pl-table-filter-criteria" ng-repeat="filter in addedFilters">
          <div class="pl-table-filter-row">
            <span class="pl-table-filter-element"><pl-input-select disable-search-filter="true" model="filter.selectedColumn" options="filter.fields" change="onSelectFilterField(filter)"></pl-input-select></span>
            <span ng-if="filter.selectedColumn !== null && getFilterTypeName(filter)!=='custom'">
              <span class="pl-table-filter-element"><pl-input-select disable-search-filter="true" model="filter.selectedOperator" options="getFilterType(filter)"></pl-input-select></span>
              <span class="pl-table-filter-element pl-table-filter-type-text" ng-if="getFilterTypeName(filter) !== 'date'">
                  <pl-input-text model="filter.text" placeholder="Enter {{filter.columns[filter.selectedColumn].label}}"></pl-input-text>
              </span>
              <span class="pl-table-filter-element pl-table-filter-type-date" ng-if="getFilterTypeName(filter)==='date'">
                  <pl-input-date-time ng-if="getFilterSubType(filter)==='birthdate'" type="date" icon-right="schedule" model="filter.text" allow-typed-input="true" format-model="YYYY-MM-DD" format-display="MM/DD/YYYY" placeholder="mm/dd/yyyy" sub-type="birthdate"></pl-input-date-time>
                  <pl-input-date-time ng-if="!getFilterSubType(filter)" type="date" icon-right="schedule" model="filter.text" allow-typed-input="true" format-model="YYYY-MM-DD" format-display="MM/DD/YYYY" placeholder="mm/dd/yyyy"></pl-input-date-time>
              </span>
            </span>
            <div class="pl-table-filter-element pl-table-filter-type-custom" ng-if="filter.selectedColumn !== null && getFilterTypeName(filter)==='custom'">
              <div class="pl-table-filter-element">
                <pl-input-select disable-search-filter="true" class="pl-table-filter-type-custom-select" model="filter.text" options="getCustomFilters(filter)" placeholder="PICK OPTION"></pl-input-select>
              </div>
            </div>
            <div ng-if="!($last && $first)" class="pl-table-filter-remove" ng-click="removeFilter(filter)">
              <figure>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 40 40">
    <path d="M27.5 10.9L20 18.4l-7.5-7.5-1.6 1.6 7.5 7.5-7.5 7.5 1.6 1.6 7.5-7.5 7.5 7.5 1.6-1.6-7.5-7.5 7.5-7.5">
</path>
    <path d="M20 1.4C9.8 1.4 1.4 9.8 1.4 20S9.8 38.6 20 38.6 38.6 30.2 38.6 20 30.2 1.4 20 1.4zm0 35.4c-9.3 0-16.8-7.5-16.8-16.8S10.7 3.2 20 3.2 36.8 10.7 36.8 20 29.3 36.8 20 36.8z">
</path>
</svg>

              </figure>
            </div>
          </div>
          <div class="pl-table-filter-row-between" ng-if="!$last">
            And
          </div>
        </div>
        <div>
          <button class="pl-button bare" ng-click="addFilterRow()" ng-disabled="!isFilterCriteriaCompleted">Add Filter</button>
        </div>
      </div>
    </div>
  </div>


  <div class="pl-table-init-loader" ng-show="ajaxInAction || !isTableInitialized">
    <div class="pl-table-loader-modal" ng-class="{'isTableInitialized': isTableInitialized}">
      <dot-loader></dot-loader>
    </div>
  </div>


  <div class="pl-table-main">

    <!-- pl-table  gets inserted here into pl-table-container -->
    <div class="pl-table-container" ng-show="((config.core.data && config.core.data.length > 0) || table.page.info().recordsDisplay > 0)"></div>

    <!-- no results message -->
    <div class="pl-table-no-results" ng-if="!ajaxInAction && ((config.core.ajax && config.pl.ajaxResponse) || !config.core.ajax) && table.page.info().recordsDisplay === 0">
      <div class="warning alert">NO RESULTS AVAILABLE</div>
    </div>
    <div class="pl-pagination-bar" ng-if="config.core.paging && table.page.info().recordsDisplay > 0">
      <span class="pl-paging-info">
        <span class="pl-table-filtering-n-results">Total: <span class="bold">{{config.getTableState().unfilteredTotalRecordCount | number}}</span>
          <span ng-show="config.getTableState().unfilteredTotalRecordCount !== table.page.info().recordsDisplay">
            <span class="bold"> | {{table.page.info().recordsDisplay | number}}</span> Filtered
          </span>
        </span>
      </span>
      <span class="pl-paging-controls">
        <button ng-disabled="table.page()===0" ng-class="{'disabled': table.page()===0}" ng-disabled="table.page() === 0" class='link pl-paging-control-button' ng-click='setPage("first")'>
          <figure>
            <?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg width="13px" height="13px" viewBox="0 0 13 13" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 39.1 (31720) - http://www.bohemiancoding.com/sketch -->
    <title>first</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Iconography" stroke="none" stroke-width="1">
        <g id="New-Icons---Gold-Release" transform="translate(-1181.000000, -341.000000)">
            <path d="M1193.42564,353.264664 L1192.85195,353.841287 C1192.6417,354.052904 1192.2972,354.052904 1192.08694,353.841287 L1186.75646,348.483809 C1186.74855,348.476424 1186.73923,348.471027 1186.73131,348.463073 L1186.15763,347.886451 C1186.02593,347.753799 1185.97676,347.568314 1186.01011,347.396464 C1186.02961,347.292501 1186.07878,347.193367 1186.15847,347.113549 L1186.73216,346.536927 C1186.73753,346.53153 1186.74403,346.527553 1186.74968,346.52244 L1192.08638,341.158713 C1192.29664,340.947096 1192.64085,340.947096 1192.85111,341.158713 L1193.42508,341.735336 C1193.63533,341.94667 1193.63533,342.292643 1193.42508,342.504261 L1188.45492,347.499574 L1193.42564,352.495739 C1193.6359,352.707357 1193.6359,353.05333 1193.42564,353.264664 Z M1183,341.998502 C1183,341.447045 1182.55614,341 1182,341 C1181.44772,341 1181,341.447489 1181,341.998502 L1181,353.001498 C1181,353.552955 1181.44386,354 1182,354 C1182.55228,354 1183,353.552511 1183,353.001498 L1183,341.998502 Z" id="first"></path>
        </g>
    </g>
</svg>

          </figure>
        </button>
        <button ng-disabled="table.page()===0" ng-class="{'disabled': table.page()===0}" class='link prev' ng-click='setPage("previous")'>
          <figure>
            <?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg width="8px" height="13px" viewBox="0 0 8 13" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 39.1 (31720) - http://www.bohemiancoding.com/sketch -->
    <title>previous</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Iconography" stroke="none" stroke-width="1">
        <g id="New-Icons---Gold-Release" transform="translate(-1208.000000, -341.000000)">
            <path d="M1215.42564,353.264664 L1214.85195,353.841287 C1214.6417,354.052904 1214.2972,354.052904 1214.08694,353.841287 L1208.75646,348.483809 C1208.74855,348.476424 1208.73923,348.471027 1208.73131,348.463073 L1208.15763,347.886451 C1208.02593,347.753799 1207.97676,347.568314 1208.01011,347.396464 C1208.02961,347.292501 1208.07878,347.193367 1208.15847,347.113549 L1208.73216,346.536927 C1208.73753,346.53153 1208.74403,346.527553 1208.74968,346.52244 L1214.08638,341.158713 C1214.29664,340.947096 1214.64085,340.947096 1214.85111,341.158713 L1215.42508,341.735336 C1215.63533,341.94667 1215.63533,342.292643 1215.42508,342.504261 L1210.45492,347.499574 L1215.42564,352.495739 C1215.6359,352.707357 1215.6359,353.05333 1215.42564,353.264664" id="previous"></path>
        </g>
    </g>
</svg>

          </figure>
        </button>
        <span class="current-page"> Page {{table.page()+1 | number}}&nbsp; of &nbsp;{{getNumPages() | number}} </span>
        <button ng-disabled="table.page()+1===getNumPages()" ng-class="{'disabled': table.page()+1===getNumPages()}" class='link next' ng-click='setPage("next")'>
          <figure>
            <?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg width="8px" height="13px" viewBox="0 0 8 13" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 39.1 (31720) - http://www.bohemiancoding.com/sketch -->
    <title>next</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Iconography" stroke="none" stroke-width="1">
        <g id="New-Icons---Gold-Release" transform="translate(-1232.000000, -341.000000)">
            <path d="M1239.42564,353.264664 L1238.85195,353.841287 C1238.6417,354.052904 1238.2972,354.052904 1238.08694,353.841287 L1232.75646,348.483809 C1232.74855,348.476424 1232.73923,348.471027 1232.73131,348.463073 L1232.15763,347.886451 C1232.02593,347.753799 1231.97676,347.568314 1232.01011,347.396464 C1232.02961,347.292501 1232.07878,347.193367 1232.15847,347.113549 L1232.73216,346.536927 C1232.73753,346.53153 1232.74403,346.527553 1232.74968,346.52244 L1238.08638,341.158713 C1238.29664,340.947096 1238.64085,340.947096 1238.85111,341.158713 L1239.42508,341.735336 C1239.63533,341.94667 1239.63533,342.292643 1239.42508,342.504261 L1234.45492,347.499574 L1239.42564,352.495739 C1239.6359,352.707357 1239.6359,353.05333 1239.42564,353.264664" id="next" transform="translate(1235.791667, 347.500000) scale(-1, 1) translate(-1235.791667, -347.500000) "></path>
        </g>
    </g>
</svg>

          </figure>
        </button>
        <button ng-disabled="table.page()+1===getNumPages()" ng-class="{'disabled': table.page()+1===getNumPages()}" class='link' ng-click='setPage("last")'>
          <figure>
            <?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg width="13px" height="13px" viewBox="0 0 13 13" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 39.1 (31720) - http://www.bohemiancoding.com/sketch -->
    <title>last</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Iconography" stroke="none" stroke-width="1">
        <g id="New-Icons---Gold-Release" transform="translate(-1253.000000, -341.000000)">
            <path d="M1265.42564,353.264664 L1264.85195,353.841287 C1264.6417,354.052904 1264.2972,354.052904 1264.08694,353.841287 L1258.75646,348.483809 C1258.74855,348.476424 1258.73923,348.471027 1258.73131,348.463073 L1258.15763,347.886451 C1258.02593,347.753799 1257.97676,347.568314 1258.01011,347.396464 C1258.02961,347.292501 1258.07878,347.193367 1258.15847,347.113549 L1258.73216,346.536927 C1258.73753,346.53153 1258.74403,346.527553 1258.74968,346.52244 L1264.08638,341.158713 C1264.29664,340.947096 1264.64085,340.947096 1264.85111,341.158713 L1265.42508,341.735336 C1265.63533,341.94667 1265.63533,342.292643 1265.42508,342.504261 L1260.45492,347.499574 L1265.42564,352.495739 C1265.6359,352.707357 1265.6359,353.05333 1265.42564,353.264664 Z M1255,341.998502 C1255,341.447045 1254.55614,341 1254,341 C1253.44772,341 1253,341.447489 1253,341.998502 L1253,353.001498 C1253,353.552955 1253.44386,354 1254,354 C1254.55228,354 1255,353.552511 1255,353.001498 L1255,341.998502 Z" id="last" transform="translate(1259.291667, 347.500000) scale(-1, 1) translate(-1259.291667, -347.500000) "></path>
        </g>
    </g>
</svg>

          </figure>
        </button>
      </span>
      <span class="pl-paging-size">
        Items per Page: &nbsp;
        <pl-input-select disable-search-filter="true" model="config.pageLength" options="[{label: '5', value:5},{label: '10', value:10},{label: '25', value:25}, {label: '50', value: 50}, {label: '100', value: 100}]" change="setPageSize()" placeholder="page size"></pl-input-select>
      </span>
    </div>
  </div>

</div>

```

**Comment by brian-mann** [comment]
Okay I see what is going on. Can you show some application code which loads in this template? This would save us some time.

I'd like to inspect the http request to see if there's a way we can detect this, other than providing a configuration option to whitelist specific urls.

**Comment by jhanink** [comment]
we associate a template location path with an angular route configuration (which binds view and logic to routes). Here is a HAR file.
[cypress-issue-257.har.zip](https://github.com/cypress-io/cypress/files/509418/cypress-issue-257.har.zip)

**Comment by jhanink** [comment]
this probably isn't too helpful, but this is where we bind our template to the angular constructs:

```
function plTableDirective($timeout, $compile, $log, $http, TableFilter, plUserPreferences) {
  return {
    restrict: 'E',
    scope: {
      config: '=', // plTableConfig object
      configValues: '=', // a literal config value object
    },
    replace: true,
    templateUrl: '/core/pl-table/src/plTable/plTable.tpl.html',
    link: ($scope, $element, $attrs) => {
      $scope.userPrefs = plUserPreferences;
```

**Comment by brian-mann** [comment]
Does this example app do the same thing as how you load your templates? I'm just looking to recreate your setup so we can write some tests around this behavior and inspect exactly what is going on.

http://airpair.github.io/demos/2014/09/T0021-airpair-angularjs-tutorial/#/

**Comment by brian-mann** [comment]
I tried that example repo but it works even with injection. Is there any way you can create a simple reduced repo demonstrating the problem?

I understand why this is happening but would like to be able to reproduce it locally. I'm not an Angular user so I'm not sure what has to be done for templates to be fetched over XHR and what exactly causes the error with the root node. So far I'm setting partials get fetched over XHR without any problems. What is the difference between a partial and a template?

I'm hoping whatever it is can be demonstrated in just a few lines of code.

**Comment by jhanink** [comment]
sure. I'll find (or make) a simple example.

There is no difference in template type, the issue crops up in usage. In the app. templates get compiled, and if we compile one of these fragments, this problem surfaces. If we merge templates first where we retain an outer tag before compile, the issue will not manifest.

I'll come back with an example tomorrow.

**Comment by jhanink** [comment]
I was able to use a workaround in our code, since angular directives have many options. I am able to ensure that our loaded template gets injected into another containing div element such that the resulting pre-compile template has a single root element.

**Comment by jhanink** [comment]
unfortunately, there are some cases where I'm not able to work around it, such as when I don't have full control over the component that I'm loading.

**Comment by brian-mann** [comment]
Fixed in [`0.17.10`](https://github.com/cypress-io/cypress/wiki/Changelog#01710-11072016).

**Comment by brian-mann** [comment]
I'm actually *assuming* this was fixed but cannot be 100% sure because I was never able to find an angular app to replicate loading the templates this way.

Can you update and verify this does indeed fix your problem?

**Comment by jhanink** [comment]
@brian-mann - thanks for the fix! yes, I tested v 0.17.10 with my existing workaround (http interceptor that removed the injected <head> content, and then after without it, and it works!
